File: SizeUpSwapperInObjc/SizeUpSwapperInObjc/WindowDetector.m
//
//  WindowDetector.m
//  SizeUpSwapperInObjc
//
//  Created by Wei Sun on 2025/8/12.
//

#import "WindowDetector.h"
#import <CoreGraphics/CoreGraphics.h>

@implementation ScreenInfo
@end

@implementation ScreenWindowPair
@end

@implementation WindowDetector

+ (NSArray<ScreenWindowPair *> *)topWindowsPerScreen {
    CFArrayRef windowList = CGWindowListCopyWindowInfo(
        kCGWindowListOptionOnScreenOnly | kCGWindowListExcludeDesktopElements,
        kCGNullWindowID
    );
    
    NSArray<NSScreen *> *screens = [NSScreen screens];
    NSMutableArray<ScreenWindowPair *> *result = [NSMutableArray array];
    
    // Initialize result with all screens
    for (NSScreen *screen in screens) {
        ScreenWindowPair *pair = [[ScreenWindowPair alloc] init];
        pair.screen = screen;
        pair.window = nil;
        [result addObject:pair];
    }
    
    if (!windowList) {
        return [result copy];
    }
    
    NSArray *windows = (__bridge NSArray *)windowList;
    
    for (NSDictionary *windowDict in windows) {
        NSNumber *layerNum = windowDict[(__bridge NSString *)kCGWindowLayer];
        if (!layerNum || [layerNum integerValue] != 0) {
            continue;
        }
        
        NSNumber *pidNum = windowDict[(__bridge NSString *)kCGWindowOwnerPID];
        NSNumber *windowIDNum = windowDict[(__bridge NSString *)kCGWindowNumber];
        NSDictionary *boundsDict = windowDict[(__bridge NSString *)kCGWindowBounds];
        
        if (!pidNum || !windowIDNum || !boundsDict) {
            continue;
        }
        
        CGRect windowBounds;
        CGRectMakeWithDictionaryRepresentation((__bridge CFDictionaryRef)boundsDict, &windowBounds);
        
        NSString *appName = windowDict[(__bridge NSString *)kCGWindowOwnerName];
        NSString *windowTitle = windowDict[(__bridge NSString *)kCGWindowName];
        
        WindowInfo *windowInfo = [[WindowInfo alloc] 
            initWithWindowID:[windowIDNum unsignedIntValue]
            ownerPID:[pidNum intValue]
            frame:windowBounds
            appName:appName
            windowTitle:windowTitle
            layer:[layerNum integerValue]];
        
        // Find which screen this window belongs to
        for (ScreenWindowPair *pair in result) {
            if (pair.window == nil && 
                NSIntersectsRect(pair.screen.frame, NSRectFromCGRect(windowBounds))) {
                pair.window = windowInfo;
                break;
            }
        }
        
        // Check if all screens have windows
        BOOL allScreensHaveWindows = YES;
        for (ScreenWindowPair *pair in result) {
            if (pair.window == nil) {
                allScreensHaveWindows = NO;
                break;
            }
        }
        
        if (allScreensHaveWindows) {
            break;
        }
    }
    
    CFRelease(windowList);
    return [result copy];
}

+ (NSArray<WindowInfo *> *)getAllWindowsInfo {
    CFArrayRef windowList = CGWindowListCopyWindowInfo(
        kCGWindowListOptionOnScreenOnly | kCGWindowListExcludeDesktopElements,
        kCGNullWindowID
    );
    
    if (!windowList) {
        return @[];
    }
    
    NSArray *windows = (__bridge NSArray *)windowList;
    NSMutableArray<WindowInfo *> *result = [NSMutableArray array];
    
    for (NSDictionary *windowDict in windows) {
        NSNumber *layerNum = windowDict[(__bridge NSString *)kCGWindowLayer];
        if (!layerNum || [layerNum integerValue] != 0) {
            continue;
        }
        
        NSNumber *pidNum = windowDict[(__bridge NSString *)kCGWindowOwnerPID];
        NSNumber *windowIDNum = windowDict[(__bridge NSString *)kCGWindowNumber];
        NSDictionary *boundsDict = windowDict[(__bridge NSString *)kCGWindowBounds];
        
        if (!pidNum || !windowIDNum || !boundsDict) {
            continue;
        }
        
        CGRect windowBounds;
        CGRectMakeWithDictionaryRepresentation((__bridge CFDictionaryRef)boundsDict, &windowBounds);
        
        NSString *appName = windowDict[(__bridge NSString *)kCGWindowOwnerName];
        NSString *windowTitle = windowDict[(__bridge NSString *)kCGWindowName];
        
        WindowInfo *windowInfo = [[WindowInfo alloc] 
            initWithWindowID:[windowIDNum unsignedIntValue]
            ownerPID:[pidNum intValue]
            frame:windowBounds
            appName:appName
            windowTitle:windowTitle
            layer:[layerNum integerValue]];
        
        [result addObject:windowInfo];
    }
    
    CFRelease(windowList);
    return [result copy];
}

+ (NSArray<ScreenInfo *> *)getScreenInfo {
    NSArray<NSScreen *> *screens = [NSScreen screens];
    NSMutableArray<ScreenInfo *> *result = [NSMutableArray array];
    
    for (NSScreen *screen in screens) {
        ScreenInfo *info = [[ScreenInfo alloc] init];
        info.screen = screen;
        info.name = screen.localizedName;
        [result addObject:info];
    }
    
    return [result copy];
}

@end

File: Sources/WindowCore/WindowDetector.swift
import Foundation
import AppKit
import CoreGraphics

public struct WindowDetector {
    
    public static func topWindowsPerScreen() -> [NSScreen: WindowInfo] {
        guard let windowInfoList = CGWindowListCopyWindowInfo(
            [.optionOnScreenOnly, .excludeDesktopElements],
            kCGNullWindowID
        ) as NSArray? else {
            return [:]
        }
        
        var result: [NSScreen: WindowInfo] = [:]
        
        for case let windowDict as NSDictionary in windowInfoList {
            guard let layer = windowDict[kCGWindowLayer as String] as? Int,
                  layer == 0,
                  let pid = windowDict[kCGWindowOwnerPID as String] as? pid_t,
                  let windowID = windowDict[kCGWindowNumber as String] as? CGWindowID,
                  let boundsDict = windowDict[kCGWindowBounds as String] as? NSDictionary,
                  let windowBounds = CGRect(dictionaryRepresentation: boundsDict) else {
                continue
            }
            
            let appName = windowDict[kCGWindowOwnerName as String] as? String
            let windowTitle = windowDict[kCGWindowName as String] as? String
            
            let windowInfo = WindowInfo(
                id: windowID,
                ownerPID: pid,
                frame: windowBounds,
                appName: appName,
                windowTitle: windowTitle,
                layer: layer
            )
            
            for screen in NSScreen.screens {
                if screen.frame.intersects(windowBounds) && result[screen] == nil {
                    result[screen] = windowInfo
                    break
                }
            }
            
            if result.count == NSScreen.screens.count {
                break
            }
        }
        
        return result
    }
    
    public static func getAllWindowsInfo() -> [WindowInfo] {
        guard let windowInfoList = CGWindowListCopyWindowInfo(
            [.optionOnScreenOnly, .excludeDesktopElements],
            kCGNullWindowID
        ) as NSArray? else {
            return []
        }
        
        var windows: [WindowInfo] = []
        
        for case let windowDict as NSDictionary in windowInfoList {
            guard let layer = windowDict[kCGWindowLayer as String] as? Int,
                  layer == 0,
                  let pid = windowDict[kCGWindowOwnerPID as String] as? pid_t,
                  let windowID = windowDict[kCGWindowNumber as String] as? CGWindowID,
                  let boundsDict = windowDict[kCGWindowBounds as String] as? NSDictionary,
                  let windowBounds = CGRect(dictionaryRepresentation: boundsDict) else {
                continue
            }
            
            let appName = windowDict[kCGWindowOwnerName as String] as? String
            let windowTitle = windowDict[kCGWindowName as String] as? String
            
            let windowInfo = WindowInfo(
                id: windowID,
                ownerPID: pid,
                frame: windowBounds,
                appName: appName,
                windowTitle: windowTitle,
                layer: layer
            )
            
            windows.append(windowInfo)
        }
        
        return windows
    }
    
    public static func getScreenInfo() -> [(screen: NSScreen, name: String)] {
        return NSScreen.screens.enumerated().map { index, screen in
            let name = screen.localizedName
            return (screen, name)
        }
    }
}